util/netevent.c

```diff
@@ -1232,6 +1232,13 @@ ssl_handshake(struct comm_point* c)
 		if(want == SSL_ERROR_WANT_READ) {
 			if(c->ssl_shake_state == comm_ssl_shake_read)
 				return 1;
+			/* According to https://www.openssl.org/docs/man1.1.1/man3/SSL_do_handshake.html
+			 * we should repeat handshake - for non blocking BIO
+			 */
+			if(c->ssl_shake_state == comm_ssl_shake_write) {
+			    comm_point_listen_for_rw(c, 0, 1);
+			    return 1;
+			}
```

Comment:
With this condition now, the code that follows turns to dead code.
I would rewrite as:
```
if(want == SSL_ERROR_WANT_READ) {
			/* According to https://www.openssl.org/docs/man1.1.1/man3/SSL_do_handshake.html
			 * we should repeat handshake - for non blocking BIO
			 */
			if(c->ssl_shake_state == comm_ssl_shake_write) {
			    comm_point_listen_for_rw(c, 0, 1);
			}
			return 1;
```

The same logic, adjusted, can also be used for the the `SSL_ERROR_WANT_WRITE` case below.

Reply:
We had a better look for this change and we solved our case (https on windows) with commit https://github.com/NLnetLabs/unbound/commit/ca67691092127aafd3c11ead622dcb659959e1bf.
Can you try that one and see if it also solves your use case?

Btw the ssl handshake is (re)tried on both read and write events until the c->ssl_shake_state reaches comm_ssl_shake_none.

Reply:
Thanks! This solution resolves also my use cases.